### ISSUE TITLE
Retuning an instance of the class being defined which provides an interface is not handled

### ISSUE THREAD

**Issue by wsanchez** [Shoobx/mypy-zope#39]
In https://github.com/twisted/klein/pull/471, I'm running into the following [error in CI](https://github.com/twisted/klein/runs/2097187159):

```console
src/klein/_dihttp.py:62:16: error: Incompatible return value type (got "RequestURL", expected "IDependencyInjector")  [return-value]
            return cls()
```

The [code in question](https://github.com/twisted/klein/blob/b46383c96ef5d0e9fbd186b3b47336feacd26bb1/src/klein/_dihttp.py#L62) is:

```python
@provider(IRequiredParameter, IDependencyInjector)
class RequestURL:
    """
    Require a hyperlink L{DecodedURL} object from a L{Requirer}.

    @since: Klein NEXT
    """

    @classmethod
    def registerInjector(
        cls,
        injectionComponents: Componentized,
        parameterName: str,
        requestLifecycle: IRequestLifecycle,
    ) -> IDependencyInjector:
        return cls()

…
```

So here, we return a `RequestURL`, which is declared as a provider of `IDependencyInjector`, so I think the above error is incorrect.

I'm guessing this is due to the fact that the definition of `RequestURL` isn't complete yet?

**Comment by wsanchez** [comment]
I think the actual issue is that `@provider` isn't handled, unlike `@implementer`.

**Comment by kedder** [comment]
Yes, mypy-zope doesn't know anything about `@provider`. That said, I think your declaration is not quite correct? If instance of `RequestURL` implements `IDependencyInjector`, as `registerInjector` suggests, shouldn't `RequestURL` be declared with `@implementer` instead?

**Comment by wsanchez** [comment]
Heh… I had to ask @glyph about this also.  `@implementer` says that instances of the class will provide the interface.  `@provider` says that the class itself does so.  (Note that the methods are all class methods.)

**Comment by wsanchez** [comment]
Super happy to be getting real typing via `mypy-zope`, by the way.  It's great.  And thanks for the quick replies to these tickets.

**Comment by kedder** [comment]
> Heh… I had to ask @glyph about this also. `@implementer` says that instances of the class will provide the interface. `@provider` says that the class itself does so. (Note that the methods are all class methods.)

Right, but `registerInjector` returns _an instance_ of `RequestURL`. So, in order for `registerInjector` return type to be correct, an instance of `RequestURL` must provide the `IDependencyInjector` . However, it does not provide that interface. Only the class itself does, according to `@provider` declaration.

So, even though  `@provider`  is currently not supported by `mypy-zope`, I suspect the type hint on `registerInjector` is not entirely correct. I'd say `registerInjector` has to return `cls`, not `cls()`. Or, alterntively, `RequestURL` must _implement_ `IDependencyInjector`, not _provide_ it.

**Comment by kedder** [comment]
> Super happy to be getting real typing via `mypy-zope`, by the way. It's great. And thanks for the quick replies to these tickets.

Thanks! I'm glad mypy exists and it was possible to marry zope with it. Interfaces become so much more meaningful with it! :)
